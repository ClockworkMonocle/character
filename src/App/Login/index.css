.login__title {
  max-width: 220px;
  margin: 0 auto;
  text-align: center;
  padding-top: 24px;
}
