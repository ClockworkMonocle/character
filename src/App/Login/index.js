import React, { Component } from 'react';
import firebaseui from 'firebaseui';
import firebase from 'firebase';
import Type from 'components/Type';
import Icon from 'components/Icon';

import './index.css';

class Login extends Component {
  componentDidMount() {
    const config = {
      signInOptions: [
        firebase.auth.GoogleAuthProvider.PROVIDER_ID,
        // firebase.auth.FacebookAuthProvider.PROVIDER_ID,
        // firebase.auth.TwitterAuthProvider.PROVIDER_ID,
        // firebase.auth.GithubAuthProvider.PROVIDER_ID,
        firebase.auth.EmailAuthProvider.PROVIDER_ID,
        // firebase.auth.PhoneAuthProvider.PROVIDER_ID,
      ],
      callbacks: {
        signInSuccess: /* istanbul ignore next */ () =>
          this.props.history.push('/app'),
      },
    };

    let fbui = firebaseui.auth.AuthUI.getInstance();
    /* istanbul ignore next */
    if (fbui) {
      /* istanbul ignore next */
      fbui.reset();
    } else {
      fbui = new firebaseui.auth.AuthUI(firebase.auth());
    }

    fbui.start('#login', config);
  }

  render() {
    return [
      <div key="login__title" className="login__title">
        <Icon name="user-secret" size={64} />
        <Type variant="display-1">Who are you?</Type>
      </div>,
      <div key="login__form" id="login" />,
    ];
  }
}

export default Login;
