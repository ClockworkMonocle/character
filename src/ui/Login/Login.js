import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import { withRouter } from 'react-router-dom';
import firebase from 'firebase';

import Message from '../components/Message';
import Icon from '../components/Icon';

import { ref } from '../../api';
import { signOut } from '../state/actions';
import { generateRandomName } from '../generateName';

const IDENTIFY = 'identify';
const SIGNUP = 'signup';
const PASSWORD = 'password';

export class Login extends Component {
  static displayName = 'Login';

  state = {
    messageType : "alert",
    disabled: false,
    isLoading: false,
    error: '',
    formType: IDENTIFY,
    email: '',
  }

  emailInput = null;
  pwdInput = null;

  componentDidMount() {
    if (!window.sessionStorage.getItem('__pocket_character_redirect_login__')) {
      this.props.dispatch(signOut());
    }
  }

  componentWillMount() {
    if (window.sessionStorage.getItem('__pocket_character_redirect_login__') === 'LOGIN') {
      this.setState({ isLoading: true });
      firebase.auth()
        .getRedirectResult()
        .then(result => result.user)
        .then(user => {
          return ref
            .child(`users/${user.uid}`)
            .once('value')
            .then(snapshot => snapshot.val())
            .then(userData => {
              if (!userData) {
                return ref
                  .child(`users/${user.uid}`)
                  .update({ createdOn: Date.now() });
              }
            })
        })
        .then(() => {
          let currentUser = firebase.auth().currentUser;

          // only update if displayName doesn't exist yet
          // pull displayName and photo from provider data;
          if (!currentUser.displayName) {
            return currentUser.updateProfile({
              photoURL: currentUser.providerData[0].photoURL,
              displayName: currentUser.providerData[0].displayName
            })
          }
        })
        .then(() => {
          window.sessionStorage.setItem('__pocket_character_redirect_login__', false);
          this.props.dispatch({ type: 'USER_LOADING_PROFILE' });
          this.props.history.push('/app');
        })
        .catch(err => {
          this.setState({ error: err.message, isLoading: false })
        });
    }
    else {
      this.setState({ isLoading: false, error: '' })
    }
  }

  handleProviderSearch = () => {
    firebase
      .auth()
      .fetchProvidersForEmail(this.emailInput.value.trim())
      .then(results => {
        let provider = results[0];

        // if results is empty, then new user, show sign up form
        // if results is federated, login via redirect
        // if results is password, show password form
        switch (provider) {
          case 'google.com':
            this.googleLogin();
            break;
          case 'password':
            this.setState({ formType: PASSWORD, email: this.emailInput.value.trim() });
            break;
          default:
            this.setState({ formType: SIGNUP, email: this.emailInput.value.trim() });
        }
      })
      .catch(error => {
        this.setState({ error: error.message });
      })
  }

  handlePasswordLogin = () => {
    let password = this.passwordInput.value.trim();
    let email = this.state.email;


    this.props.dispatch({ type: 'USER_LOADING_PROFILE' });
    this.setState({ disabled: true });
    firebase.auth()
      .signInWithEmailAndPassword(email, password)
      .then(() => {
        this.props.history.push("/app");
      })
      .catch(error => {
        this.setState({ disabled: false, error: error.message });
      });
  }

  googleLogin = () => {
    if (this.state.isLoading || this.state.disabled) return;

    this.setState({ disabled: true })

    sessionStorage.setItem('__pocket_character_redirect_login__', 'LOGIN');
    let provider = new firebase.auth.GoogleAuthProvider();
    provider.addScope('email');

    firebase.auth()
      .signInWithRedirect(provider)
      .catch(err => {
        this.setState({ err: err.message })
      });
  }

  handleAccountCreate = () => {
    let email = this.state.email;
    let pwd = this.pwdInput.value.trim();
    let confirmPwd = this.confirmPwdInput.value.trim();

    if (pwd !== confirmPwd) {
      this.setState({ error: 'Passwords must match!' })
      return;
    }

    this.props.dispatch({ type: 'USER_LOADING_PROFILE' });
    firebase.auth()
      .createUserWithEmailAndPassword(email, pwd)
      .then(user => {
        ref
          .child(`users/${user.uid}`)
          .update({ createdOn: Date.now() })
          .then(() => {
            return user.updateProfile({
              displayName: generateRandomName(),
              photoURL: '',
            })
          })
          .then(() => {
            this.props.history.push('/app');
          })
          .catch(error => {
            this.setState({ error: error.message });
          })
      })
      .catch(error => {
        this.setState({ error: error.message });
      })
  }

  renderLandingForm = () => {
    return <div className='login-input-group'>
      <Message type={this.state.messageType} message={this.state.error} />
      <input
        className='login-input'
        type='email'
        placeholder='email address'
        ref={ref => this.emailInput = ref}
      />
      <button
        className='login-btn'
        onClick={this.handleProviderSearch}
        disabled={this.state.isLoading || this.state.disabled}
      >{this.state.isLoading ? 'Authenticating...' : 'Sign In'}</button>
    </div>
  }

  renderSignUpForm = () => {
    return <div className="login-input-group">
      <Message type={this.state.messageType} message={this.state.error} />
      <p className='login-email-disabled text-center'>{this.state.email}</p>
      <hr />
      <input
        className='login-input'
        type='password'
        placeholder='new password'
        ref={ref => this.pwdInput = ref}
      />
      <input
        className='login-input'
        type='password'
        placeholder='confirm password'
        ref={ref => this.confirmPwdInput = ref}
      />
      <button
        className='login-btn'
        onClick={this.handleAccountCreate}
        disabled={this.state.isLoading || this.state.disabled}
      >{this.state.isLoading || this.state.disabled ? 'Creating...' : 'Create Account'}</button>
      <hr />
      <div className='mt3'>
        <div
          className='login-google'
          disabled={this.state.isLoading || this.state.disabled}
          onClick={this.googleLogin}
        ></div>
      </div>
    </div>
  }

  renderPasswordForm = () => {
    return <div className='login-input-group'>
      <Message type={this.state.messageType} message={this.state.error} />
      <p className='login-email-disabled text-center'>{this.state.email}</p>
      <input
        className='login-input'
        type='password'
        placeholder='password'
        ref={ref => this.passwordInput = ref}
      />
      <button
        className='login-btn'
        onClick={this.handlePasswordLogin}
        disabled={this.state.isLoading || this.state.disabled}
      >{this.state.isLoading || this.state.disabled ? 'Authenticating...' : 'Sign In'}</button>
    </div>
  }

  renderForm = () => {
    switch (this.state.formType) {
      case IDENTIFY:
        return this.renderLandingForm();
      case SIGNUP:
        return this.renderSignUpForm();
      case PASSWORD:
        return this.renderPasswordForm();
    }
  }

  render() {
    return (
      <div className="login-page">
        <div className="login-content">
          <h1 className="login-title ml3"><Icon icon='fa fa-user-secret'/></h1>
          <p>Who are you?</p>
          { this.renderForm() }
          <p className="login-subtext"><small>&copy;2016 - ClockworkMonocle</small></p>
        </div>
      </div>
    );
  }
}

export default withRouter(Login);
