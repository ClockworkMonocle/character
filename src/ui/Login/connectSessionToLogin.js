import React, { Component } from 'react';
import Router from '../router/Router';
import { ROUTE_PROFILE } from '../routes';
import firebase from 'firebase';
import { ref } from '../../api';

export function connectSessionToLogin(Login, session) {
  return class ConnectedLoginSession extends Component {
    static displayName = 'ConnectedLoginSession';

    state = {
      isLoading: false,
      error: '',
    }

    componentWillMount() {
      if (session.getItem('__pocket_character_redirect_login__') === 'LOGIN') {
        this.setState({ isLoading: true });
        firebase.auth()
          .getRedirectResult()
          .then(result => result.user)
          .then(user => {
            return ref
              .child(`users/${user.uid}`)
              .once('value')
              .then(snapshot => snapshot.val())
              .then(userData => {
                if (!userData) {
                  return ref
                    .child(`users/${user.uid}`)
                    .update({ createdOn: Date.now() });
                }
              })
          })
          .then(() => {
            let currentUser = firebase.auth().currentUser;

            // only update if displayName doesn't exist yet
            if (!currentUser.displayName) {
              return currentUser.updateProfile({
                photoURL: currentUser.providerData[0].photoURL,
                displayName: currentUser.providerData[0].displayName
              })
            }
          })
          .then(() => {
            session.setItem('__pocket_character_redirect_login__', false);
            Router.nav(ROUTE_PROFILE);
          })
          .catch(err => {
            this.setState({ error: err.message, isLoading: false })
          })
      }
      else {
        this.setState({ isLoading: false, error: '' })
      }
    }

    render() {
      return <Login
        {...this.props}
        isLoading={this.state.isLoading}
        error={this.state.error}
      />
    }
  }
}
