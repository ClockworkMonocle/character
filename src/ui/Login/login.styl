.login-page
  width 100%

.login-content
  text-align center
  padding $base-padding
  width 100%

.login-title
  text-align center
  font-weight bold

.login-subtext
  color #bbb
  margin-top 15px

.login-input-group
  width 33.33333333333%
  margin 0 auto
  @media screen and (max-width: 480px)
    width 100%

.login-input
  padding $base-padding-sm

.login-btn
  background-color lighten(#3c763d, 17%)
  color #fff
  &:active:enabled
    background-color darken(#3c763d, 10%)
  &:disabled
    background-color darken(#3c763d, 45%)
    
.login-google
  display block
  width 200px
  height 50px
  margin 0 auto
  background url('./imgs/btn_google_signin_light_normal_web@2x.png') no-repeat
  background-size 191px 46px
  
  &:active
    background-image url('./imgs/btn_google_signin_light_pressed_web@2x.png')
  &:focus
    background-image url('./imgs/btn_google_signin_light_focus_web@2x.png')
  &[disabled]
    background-image url('./imgs/btn_google_signin_light_disabled_web@2x.png')
  
.login-input,
.login-btn
  display block
  appearance none
  border 1px solid #ddd
  border-radius $base-border-radius
  width 200px
  height 50px
  margin 10px auto
  