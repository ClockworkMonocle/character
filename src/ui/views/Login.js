"use strict";

var React = require('react/addons');
var Router = require('../router/Router');
var Message = require('../components/Message');
var db = require('../../api');

module.exports = React.createClass({
  displayName : "Login",


  getInitialState : function() {
    return ({
      btnVal : "Log in",
      message : "",
      messageType : "alert",
      disabled : false
    })
  },


  componentWillMount : function() {
    var email;
    var newuser;

    if (db.ref.getAuth()) {
      Router.nav("/user");
    }
    else if (this.props.token) {
      email = localStorage.getItem("__cm_character_app_email__");
      newuser = localStorage.getItem("__cm_character_app_new_user__");
      this.setState({ btnVal : "Authenticating...", disabled : true });

      db.auth(email, this.props.token).then((auth) => {
        if (newuser) {
          var query = "/users/" + auth.uid;
          var user = {};

          user['profile_name'] = email.split("@")[0];
          user['signup_date'] = +new Date;

          return db.update(query, user);
        }
      }).then(() => {
        localStorage.removeItem("__cm_character_app_email__");
        localStorage.removeItem("__cm_character_app_new_user__");
        Router.nav('/user');
      }).catch((err) => {
        this.setState({ message : err.message, messageType : "alert", btnVal : "Log in", disabled : false });
      })
    }
  },


  clearMessage : function() {
    this.setState({ message : "" });
  },


  // regex from : http://stackoverflow.com/a/46181/3780697
  validateEmail : function(email) {
    var regx = /^([\w-]+(?:\.[\w-]+)*)@((?:[\w-]+\.)*\w[\w-]{0,66})\.([a-z]{2,6}(?:\.[a-z]{2})?)$/i;

    return regx.test(email);
  },


  generatePass : function() {
    var chars = "0123456789abcdefghijklmnopqrstuvwxyz-ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    var pass = "";

    for (var i = 0; i < 32; i++) {
      pass += chars[Math.floor(Math.random() * chars.length)];
    }

    return pass;
  },


  handleSubmit : function() {
    var email = React.findDOMNode(this.refs['input-login']).value;
    var valid = this.validateEmail(email);
    var pass = this.generatePass();

    if (!valid) {
      this.setState({ message : "Invalid email address format!" });
      return;
    }

    localStorage.setItem("__cm_character_app_email__", email);
    this.setState({ btnVal : "Sending...", disabled : true });

    // assume a new user;
    db.create(email, pass).then((auth) => {

      localStorage.setItem("__cm_character_app_new_user__", "new_user_times_yeah");
      return db.token(email);
    }).catch((err) => {
      if (err.code === "EMAIL_TAKEN") {
        return db.token(email);
      }

      throw new Error(err);
    }).then((msg) => {
      this.setState({ btnVal : "Sent", message : "Check your email!", messageType : "success" });
    }).catch((err) => {
      this.setState({ btnVal : "Submit", disabled : false, message : err.message, messageType : "alert" });
      console.error(err);
    });
  },


  render : function() {

    return (
      <div className="login-container">
        <h1 className="login-title">Authenticate</h1>
        <Message type={this.state.messageType} message={this.state.message} />
        <input ref="input-login" type="text" className="login-input-email-address" onFocus={this.clearMessage} placeholder="Email Address" />
        <button disabled={this.state.disabled} onClick={this.handleSubmit}>{this.state.btnVal}</button>
      </div>
    );
  }
})