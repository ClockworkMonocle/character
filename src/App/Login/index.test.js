import React from 'react';
import { shallow } from 'enzyme';
import Login from './index';

jest.mock('firebase');
jest.mock('firebaseui');

test('<Login /> renders', () => {
  expect(shallow(<Login />)).toMatchSnapshot();
});
