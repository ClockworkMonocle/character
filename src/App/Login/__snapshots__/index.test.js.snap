// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> renders 1`] = `
Array [
  <div
    className="login__title"
    key="login__title"
/>,
  <div
    id="login"
    key="login__form"
/>,
]
`;
